### PR TITLE
Adds DAG and task for adding 979 marc tags

### LINKED ISSUE #1185: Create a function to add a MARC tag to data in FOLIO
This function should be written so that it can be used by different DAGs that will have a task to "add some data to a MARC record" in folio. The input for this will be an instance UUID and the data to be added (probably a list of fields, such as `[{'979': {'ind1': ' ', 'ind2': ' ', 'subfields': [{'f': 'ABBOTT'}, {'b': 'druid:ws066yy0421'}, {'c': 'ws066yy0421_00_0001.jp2'}, {'d': 'The The Donald P. Abbott Fund for Marine Invertebrates'}]}}]`). We will use this function to add 979 bookplate data to a MARC record. We can look at the [OCLCWrapperAPI Class](https://github.com/sul-dlss/libsys-airflow/blob/d6b6f1c2b7b0db4b231520edd92ef814bccd3932/libsys_airflow/plugins/data_exports/oclc_api.py#L75) for an idea of the endpoints to use. (And maybe refactor OCLCWrapperAPI to use the new function.) There is also some information in the [instances paid on bookplate funds notebook](https://github.com/shelleydoljack/notebook/blob/main/digital_bookplates/instances_paid_on_bookplate_funds.ipynb) (see Get the MARC for each instance to do a PUT to add 979). We can query `change-manager/parsedRecords` to get the MARC that needs the tag data added. We also need to query `inventory/instances` to get the `_version` field to PUT to `change-manager/parsedRecords`. The function should account for overwriting existing fields in the record (if the incoming data is the same or adding the field to the MARC record. If there are errors in updating the record, these should be caught in an exception. If the error is due to optimistic locking, we should capture that for a retry after getting an updated `_version` from inventory.

### LINKED ISSUE #1186: Create a DAG and task that starts the add 979 fields DAG
The input for the add 979 fields DAG will be a list of instance UUIDs and the associated bookplate fund data. The task should get the bookplate data from the bookplates table and contruct a 979 tag with the fund name in subfield f, druid in subfield b, image filename in subfield c, and title in subfield d, such as:
`{'979': {'ind1': ' ', 'ind2': ' ', 'subfields': [{'f': 'ABBOTT'}, {'b': 'druid:ws066yy0421'}, {'c': 'ws066yy0421_00_0001.jp2'}, {'d': 'The The Donald P. Abbott Fund for Marine Invertebrates'}]}}`
The instance UUID and this data should be passed to the function to add a tag (described in #1185 ).

### LINKED ISSUE #1280: Fix broken digital_bookplate_979 DAG
Maybe this is fixed in the work Josh has been doing for the add a tag function. Latest deployment of libsys-airflow to dev shows the following for broken DAG:
```
Broken DAG: [/opt/airflow/libsys_airflow/dags/digital_bookplates/digital_bookplate_979.py]
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/models/taskmixin.py", line 262, in set_downstream
    self._set_relatives(task_or_task_list, upstream=False, edge_modifier=edge_modifier)
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/models/taskmixin.py", line 214, in _set_relatives
    task_object.update_relative(self, not upstream, edge_modifier=edge_modifier)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'update_relative'
```

### PR DESCRIPTION
**Pull request by jgreben**
Fixes #1185 
Fixes #1186
Fixes #1280 

Also adds shared utils class for adding a unique marc tag

### CODE REVIEW

**Review by jermnelson — APPROVED** [review]

File: libsys_airflow/plugins/shared/utils.py [hunk 1]
```diff
@@ -1,5 +1,114 @@
...
+                f"Failed to update FOLIO for Instance {instance_id} with SRS {srs_uuid}"
+            )
+            return False
+        return True
+
+    def __marc_json_with_new_tags__(self, marc_json: dict, marc_instances_tags: dict):
+        reader = pymarc.reader.JSONReader(json.dumps(marc_json))
+
+        for tag_name, indicator_subfields in marc_instances_tags.items():
+            for sfs in indicator_subfields['subfields']:
```
[the remark this answers is not in the document]
Comment (jgreben) [reply libsys_airflow/plugins/shared/utils.py:64]: Which is better, having the nested for loops, or passing the args down to different methods? If this is more what you had in mind, I'm not sure if it is actually easier to reason (for a human, if that is what you meant):
```
def __marc_json_with_new_tags__(self, marc_json: dict, marc_instances_tags: dict):
        for tag_name, indicator_subfields in marc_instances_tags.items():
            record = self.__create_new_marc_json__(indicator_subfields, tag_name, marc_json)

        return record.as_json()
    
    def __create_new_marc_json__(self, indicator_subfields, tag_name, marc_json):
        reader = pymarc.reader.JSONReader(json.dumps(marc_json))
        for subfields in indicator_subfields['subfields']:
            record = self.__construct_new_tag_record__(tag_name, subfields, reader)

        return record
    
    def __construct_new_tag_record__(self, tag_name, subfields, record_reader):
        for sf_code, sf_val in subfields.items():
            new_tag = pymarc.Field(
                tag=tag_name,
                indicators=[indicator_subfields['ind1'], indicator_subfields['ind2']],  # type: ignore
                subfields=[pymarc.Subfield(code=sf_code, value=sf_val)],
            )
            for record in record_reader:
                existing_tags = record.get_fields(tag_name)
                if self.__tag_is_unique__(existing_tags, new_tag):
                    record.add_field(new_tag)

        return record
```

File: tests/digital_bookplates/test_bookplates_funds.py [hunk 2]
```diff
@@ -212,3 +213,31 @@ def test_new_bookplate_funds_polines(
...
+        ],
+        "06220dd4-7d6e-4e5b-986d-5fca21d856ca": [
+            {
+                "druid": "gc698jf6425",
+                "fund_name": "RHOADES",
+                "image_filename": "gc698jf6425_00_0001.jp2",
+                "title": "John Skylstead and Carmel Cole Rhoades Fund for California History and the History of the North American West",
+            },
+        ],
+    }
```
Comment (shelleydoljack) [inline tests/digital_bookplates/test_bookplates_funds.py:236]: Maybe this is where you can test for when the fund_name is None.

File: libsys_airflow/dags/digital_bookplates/digital_bookplate_979.py [hunk 3]
```diff
@@ -30,23 +31,20 @@ def digital_bookplate_979():
 
     end = EmptyOperator(task_id="end")
 
-    instances_for_druids = launch_add_979_fields_task()
+    druids_for_instance_id = launch_add_979_fields_task()
```
Comment (shelleydoljack) [inline libsys_airflow/dags/digital_bookplates/digital_bookplate_979.py:34]: I'm not sure why this task is called here. Wouldn't the add_979_marc_tags just take the DAG params through kwargs? The launch_add_979_fields_task is called in the [digital_bookplate_instances DAG](https://github.com/sul-dlss/libsys-airflow/blob/da15bc128c0321f4ecf20ff722b54c825b1d68a3/libsys_airflow/dags/digital_bookplates/digital_bookplate_instances.py#L60). This is likely something to address in a follow-on PR once ticket #1238 if done.

**Review by shelleydoljack — APPROVED** [review]

**Review by shelleydoljack — CHANGES_REQUESTED** [review]
We need to change what goes in 979 subfield f when fund_name is None.

File: libsys_airflow/plugins/digital_bookplates/bookplates.py [hunk 4]
```diff
@@ -113,4 +126,21 @@ def construct_979_marc_tags(druid_instances: dict) -> dict:
...
+
+    marc_instances_tags: dict = {'979': []}
+    for _instance_uuid, druids in druid_instances.items():
+        for tag_data in druids:
+            marc_instances_tags['979'].append(
+                {
+                    'ind1': ' ',
+                    'ind2': ' ',
+                    'subfields': [
+                        {'f': tag_data['fund_name']},
```
[the remark this answers is not in the document]
Comment (jgreben) [reply libsys_airflow/plugins/digital_bookplates/bookplates.py:138]: Do you mean put the druid as fund_name in _this_ task function id it is `None`?

File: libsys_airflow/dags/digital_bookplates/digital_bookplate_979.py (hunk 3 above)
Comment (jgreben) [reply libsys_airflow/dags/digital_bookplates/digital_bookplate_979.py:34]: I think part of https://github.com/sul-dlss/libsys-airflow/issues/1238 should be to change this as needed, and remove the associated task if needed.

**Review by shelleydoljack — APPROVED** [review]
👍  and 🤞

**Pull request merged**